### ISSUE TITLE
Boundary detection vs type prediction

### ISSUE THREAD

**Issue by isabelleaugenstein** [clab/stack-lstm-ner#12]
I managed to train and test the model on a keyphrase identification + classification task (https://scienceie.github.io/), however I'm not sure I understand the results I get. It seems to me that the model tries to assign a type for each token given that it has the information that the token has a  non-O gold tag. To test this, I gave the model some artificial test data. I get, in two separate decoding runs, with the same trained model (output of attach_prediction.py):

=============
It O O
is O O
not O O
test B-Material I-Material
test I-Material I-Material
, O O
but B-Process I-Task
that I-Process I-Task
end O O
. O O

===============
It B-Task I-Process
is I-Task I-Process
not I-Task I-Process
test O O
test O O
, B-Process I-Process
but O O
that O O
end B-Material I-Material
. I-Material I-Material

==============

(the first column is the token, second column the gold tag, third column the predicted tag)

Is this what the model should do or am I doing something wrong? If I wanted the predicted tag for each token to be the most likely one, even if it is O, could I achieve this with your model?

**Comment by miguelballesteros** [comment]
Hi Isabelle,

I'm not sure if I understand the issue. The model is going to predict a tag (a tag can be O or whatever you have in your data). 

How did you train the model?
How is the data?

thanks,
Miguel

**Comment by isabelleaugenstein** [comment]
The issue is that, for both outputs above, I use the same model, but get different results.
The predictions of the first run completely differ from the predictions of the second run. Notably, all, and only, tokens which have a non-O tag as a gold label get assigned a non-O label as predictions.

I trained the model using the instructions you posted, substituting the training and test data. I converted the data to ConLL format, then converted it to the format your code expects using the appropriate scripts you provide.

I.e. the training and test data (in the original ConLL format before converting it) looks like this:

Nuclear B-Task
theory I-Task
devoted O
major O
efforts O
(...)

**Comment by miguelballesteros** [comment]
It is not possible to get different predictions with the same model. Once the model is trained this is fully deterministic. Just to double check, are you sure that you stopped training? (it should be stopped manually)

If I understand correctly, the system (once it is properly trained) should be able to predict what you are looking for.

I don't know how big your data is, whether you used a development set, etc.

**Comment by isabelleaugenstein** [comment]
I stopped training and I gave it the training set as a dev set (because the code seemed to require a dev set).

Just to double check, is the way to get predictions

./lstm-parse -T train.parser -d test.parser --hidden_dim 100 --lstm_input_dim 100 -w sskip.100.vectors --pretrained_dim 100 --rel_dim 20 --action_dim 20 --input_dim 100 -m latest_model -S > output.txt
python attach_prediction.py -p output.txt -t test.txt -o evaloutput.txt

where "latest_model" automatically points to the latest model, "train.parser" and "test.parser" are the training and test, respectively, converted with "convert-conll2trans.pl" followed by "conll2parser.py" and the "test.txt" is the ConLL-formatted original input data?
And should "evaloutput.txt" then contain the results in the format "token\tgold\tpredicted"?

**Comment by miguelballesteros** [comment]
Hi Isabelle,

this seems all correct yes.

What is the score in the development set when you stopped training?  (you can see it in the log file or predicting the dev set)
how long did it take until you stopped it?

can you try to predict the dev set and see whether it makes sense?

**Comment by isabelleaugenstein** [comment]
I trained for 30 epochs, then stopped as it reached good scores on the training set. This is what I get when I run the model on the training set:

0class Precision 0.987508 Recall 0.985922 F1 0.986714
1class Precision 0.982386 Recall 0.978718 F1 0.980549
2class Precision 0.974758 Recall 0.986145 F1 0.980419
F1 >> 0.982561

Which makes sense, it should be a good fit on the training data.

The gold labels + predictions on train looks as follows:

Nuclear B-Task I-Task
theory I-Task I-Task
devoted O O
major O O
efforts O O
since O O

But the same thing happens as a described above, if I change the gold labels in the data I evaluate on (in this case the training data, but note again that I use the same trained model for decoding), I get different predictions:

Nuclear O O
theory O I-Task
devoted I-Task I-Task
major I-Task I-Task
efforts I-Task I-Task
since I-Task I-Task

I'm sure I'm missing something here, but can't spot what it is.

**Comment by miguelballesteros** [comment]
Hi Isabelle, 

can you send me your output and gold file in an email? I will try to run the script and see whether I find what is going on.

**Comment by miguelballesteros** [comment]
It was a misunderstanding with the data format.
